concordium-smart-contract-testing/src/impls.rs

```diff
@@ -0,0 +1,962 @@
+    }
+
+    /// Returns the balance of an account if it exists.
+    pub fn account_balance(&self, address: AccountAddress) -> Option<AccountBalance> {
+        self.accounts.get(&address).map(|ai| ai.balance)
+    }
+
+    /// Returns the available balance of an account if it exists.
+    pub fn account_balance_available(&self, address: AccountAddress) -> Option<Amount> {
+        self.accounts.get(&address).map(|ai| ai.balance.available())
+    }
+
+    /// Returns the balance of an contract if it exists.
+    pub fn contract_balance(&self, address: ContractAddress) -> Option<Amount> {
+        self.contracts.get(&address).map(|ci| ci.self_balance)
+    }
+
+    /// Helper method for looking up part of the state of a smart contract,
+    /// which is a trie.
+    pub fn contract_state_lookup(&self, address: ContractAddress, key: &[u8]) -> Option<Vec<u8>> {
+        let mut loader = v1::trie::Loader::new(&[][..]);
+        self.contracts.get(&address)?.state.lookup(&mut loader, key)
+    }
+
+    /// Return a clone of the [`ContractModule`] (which has an `Arc` around the
+    /// artifact).
+    fn contract_module(
+        &self,
+        module_ref: ModuleReference,
+    ) -> Result<ContractModule, ModuleMissing> {
+        let module = self
+            .modules
+            .get(&module_ref)
+            .ok_or(ModuleMissing(module_ref))?;
+        Ok(module.clone())
+    }
+
+    /// Returns an immutable reference to an [`Account`].
+    pub fn account(&self, address: AccountAddress) -> Result<&Account, AccountMissing> {
+        self.accounts.get(&address).ok_or(AccountMissing(address))
+    }
+
+    /// Returns a mutable reference to [`Account`].
+    fn account_mut(&mut self, address: AccountAddress) -> Result<&mut Account, AccountMissing> {
+        self.accounts
+            .get_mut(&address)
+            .ok_or(AccountMissing(address))
+    }
+
+    /// Check whether an [`Account`] exists.
+    pub fn account_exists(&self, address: AccountAddress) -> bool {
+        self.accounts.contains_key(&address)
+    }
+
+    /// Check whether a [`Contract`] exists.
+    pub fn contract_exists(&self, address: ContractAddress) -> bool {
+        self.contracts.contains_key(&address)
+    }
+
+    /// Check whether the [`Address`] exists.
+    ///
+    /// I.e. if it is an account, whether the account exists,
+    /// and if it is a contract, whether the contract exists.
+    fn address_exists(&self, address: Address) -> bool {
+        match address {
+            Address::Account(acc) => self.account_exists(acc),
+            Address::Contract(contr) => self.contract_exists(contr),
+        }
+    }
+
+    /// Convert the [`InvokeEntrypointResult`] to a contract success or error.
+    ///
+    /// The `energy_for_state_increase` is only used if the result was a
+    /// success.
+    ///
+    /// The `state_changed` should refer to whether the state of the top-level
+    /// contract invoked has changed.
```

Comment:
Yes. That is why it lives in the `impls.rs` instead of `invocation/impls.rs`. The function is only used in two locations: `contract_update` and `contract_invoke`.